File: src/test/java/com/polygloat/controllers/TranslationControllerTest.java
package com.polygloat.controllers;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.polygloat.dtos.response.SourceResponseDTO;
import com.polygloat.dtos.response.ViewDataResponse;
import com.polygloat.dtos.response.translations_view.ResponseParams;
import com.polygloat.helpers.JsonHelper;
import com.polygloat.model.Repository;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.ResultActions;
import org.testng.annotations.Test;

import java.util.LinkedHashSet;
import java.util.Map;

import static com.polygloat.controllers.LoggedRequestFactory.loggedGet;
import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
public class TranslationControllerTest extends SignedInControllerTest {

    @Test
    @DirtiesContext(methodMode = DirtiesContext.MethodMode.BEFORE_METHOD)
    void getViewDataSearch() throws Exception {
        Repository app = dbPopulator.populate(generateUniqueString());

        String searchString = "This";

        ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> response = performValidViewRequest(app, "?search=" + searchString);
        assertThat(response.getData().size()).isGreaterThan(0);
        assertSearch(response, searchString);
    }

    @Test
    @DirtiesContext(methodMode = DirtiesContext.MethodMode.BEFORE_METHOD)
    void getViewDataQueryLanguages() throws Exception {
        Repository repository = dbPopulator.populate(generateUniqueString());

        ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> response = performValidViewRequest(repository, "?languages=en");

        assertThat(response.getData().size()).isGreaterThan(8);

        for (SourceResponseDTO item : response.getData()) {
            assertThat(item.getTranslations()).doesNotContainKeys("de");
        }

        performGetDataForView(repository.getId(), "?languages=langNotExists").andExpect(status().isNotFound());
        
        //with starting emtpy string
        ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> response2 = performValidViewRequest(repository, "?languages=,en,de");

        //with trailing empty string
        ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> response3 = performValidViewRequest(repository, "?languages=,en,de,");

        //with same language multiple times
        ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> response4 = performValidViewRequest(repository, "?languages=,en,en,,");
    }

    private ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> performValidViewRequest(Repository repository, String queryString) throws Exception {
        MvcResult mvcResult = performGetDataForView(repository.getId(), queryString).andExpect(status().isOk()).andReturn();

        ObjectMapper mapper = new ObjectMapper();

        return mapper.readValue(mvcResult.getResponse().getContentAsString(),
                new TypeReference<>() {
                });
    }

    @Test
    void getViewDataQueryPagination() throws Exception {
        Repository repository = dbPopulator.populate(generateUniqueString());

        int limit = 5;

        ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> response = performValidViewRequest(repository, String.format("?limit=%d", limit));

        assertThat(response.getData().size()).isEqualTo(limit);
        assertThat(response.getPaginationMeta().getAllCount()).isEqualTo(12);
        assertThat(response.getPaginationMeta().getOffset()).isEqualTo(0);


        int offset = 3;

        ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> responseOffset = performValidViewRequest(repository, String.format("?limit=%d&offset=%d", limit, offset));

        assertThat(responseOffset.getData().size()).isEqualTo(limit);
        assertThat(responseOffset.getPaginationMeta().getOffset()).isEqualTo(offset);

        response.getData().stream().limit(offset).forEach(i -> assertThat(responseOffset.getData()).doesNotContain(i));

        response.getData().stream().skip(offset).forEach(i -> {
            assertThat(responseOffset.getData()).contains(i);
        });
    }

    @Test
    void getViewDataMetadata() throws Exception {
        Repository repository = dbPopulator.populate(generateUniqueString());
        int limit = 5;
        ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> response = performValidViewRequest(repository, String.format("?limit=%d", limit));

        assertThat(response.getParams().getLanguages()).contains("en", "de");
    }

    private static void assertSearch(ViewDataResponse<LinkedHashSet<SourceResponseDTO>, ResponseParams> response, String searchString) {
        for (SourceResponseDTO item : response.getData()) {
            assertThat(JsonHelper.asJsonString(item)).contains(searchString);
        }
    }

    @Test
    void getTranslations() throws Exception {
        Repository repository = dbPopulator.populate(generateUniqueString());

        MvcResult mvcResult = mvc.perform(
                loggedGet("/api/repository/" + repository.getId().toString() + "/translations/en,de")
                        .contentType(MediaType.APPLICATION_JSON)).andExpect(status().isOk()).andReturn();

        ObjectMapper mapper = new ObjectMapper();

        Map<String, Object> result = mapper.readValue(mvcResult.getResponse().getContentAsString(), Map.class);
        assertThat(result).containsKeys("en", "de");
    }

   /* @Test
    @Rollback
    void getSourceTranslations() throws Exception {
        dbPopulator.populate("app4");

        Repository repository = repositoryService.findByName("app4", userAccount).orElseThrow(NotFoundException::new);

        MvcResult mvcResult = mvc.perform(
                loggedGet("/api/repository/" + repository.getId().toString() +
                        "/translations/source/sampleApp.this_is_standard_text_somewhere_in_dom")
                        .contentType(MediaType.APPLICATION_JSON)).andExpect(status().isOk()).andReturn();

        ObjectMapper mapper = new ObjectMapper();

        Map<String, Object> result = mapper.readValue(mvcResult.getResponse().getContentAsString(), Map.class);
        assertThat(result).containsKeys("en", "de");
    }*/

    private ResultActions performGetDataForView(Long repositoryId, String queryString) throws Exception {
        return mvc.perform(
                loggedGet("/api/repository/" + repositoryId + "/translations/view" + queryString)
                        .contentType(MediaType.APPLICATION_JSON));
    }

}

File: backend/app/src/test/kotlin/io/tolgee/controllers/TranslationControllerTest.kt
package io.tolgee.controllers

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import io.tolgee.development.testDataBuilder.data.LanguagePermissionsTestData
import io.tolgee.dtos.PathDTO
import io.tolgee.dtos.request.translation.SetTranslationsWithKeyDto
import io.tolgee.dtos.response.KeyWithTranslationsResponseDto
import io.tolgee.dtos.response.ViewDataResponse
import io.tolgee.dtos.response.translations_view.ResponseParams
import io.tolgee.fixtures.andIsForbidden
import io.tolgee.fixtures.generateUniqueString
import io.tolgee.fixtures.mapResponseTo
import io.tolgee.model.Project
import io.tolgee.model.enums.ApiScope
import io.tolgee.testing.annotations.ProjectApiKeyAuthTestMethod
import io.tolgee.testing.assertions.Assertions.assertThat
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.Test
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.web.servlet.ResultActions
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

@SpringBootTest
@AutoConfigureMockMvc
class TranslationControllerTest :
  ProjectAuthControllerTest() {

  @Test
  fun getViewDataSearch() {
    val app = dbPopulator.populate(generateUniqueString())
    val searchString = "This"
    val response = performValidViewRequest(app, "?search=$searchString")
    Assertions.assertThat(response.data.size).isPositive
    assertThat(jacksonObjectMapper().writeValueAsString(response)).contains(searchString)
  }

  @Test
  fun getViewDataQueryLanguages() {
    val project = dbPopulator.populate(generateUniqueString())
    val response = performValidViewRequest(project, "?languages=en")
    Assertions.assertThat(response.data.size).isGreaterThan(8)
    for ((_, _, translations) in response.data) {
      Assertions.assertThat(translations).doesNotContainKeys("de")
    }
    performGetDataForView(project.id, "?languages=langNotExists").andExpect(status().isNotFound)

    val result = performValidViewRequest(project, "?languages=en,en")
    assertThat(result.data).isNotEmpty
  }

  private fun performValidViewRequest(project: Project, queryString: String):
    ViewDataResponse<LinkedHashSet<KeyWithTranslationsResponseDto>, ResponseParams> {
    val mvcResult = performGetDataForView(project.id, queryString).andExpect(status().isOk).andReturn()
    return mvcResult.mapResponseTo()
  }

  @Test
  fun getViewDataQueryPagination() {
    val project = dbPopulator.populate(generateUniqueString())
    val limit = 5
    val response = performValidViewRequest(project, String.format("?limit=%d", limit))
    Assertions.assertThat(response.data.size).isEqualTo(limit)
    Assertions.assertThat(response.paginationMeta.allCount).isEqualTo(12)
    Assertions.assertThat(response.paginationMeta.offset).isZero
    val offset = 3
    val responseOffset = performValidViewRequest(project, String.format("?limit=%d&offset=%d", limit, offset))
    Assertions.assertThat(responseOffset.data.size).isEqualTo(limit)
    Assertions.assertThat(responseOffset.paginationMeta.offset).isEqualTo(offset)
    response.data.stream().limit(offset.toLong())
      .forEach { i: KeyWithTranslationsResponseDto ->
        Assertions.assertThat(responseOffset.data).doesNotContain(i)
      }
    response.data.stream().skip(offset.toLong())
      .forEach { i: KeyWithTranslationsResponseDto ->
        Assertions.assertThat(responseOffset.data).contains(i)
      }
  }

  @Test
  fun getViewDataMetadata() {
    val project = dbPopulator.populate(generateUniqueString())
    val limit = 5
    val response = performValidViewRequest(project, String.format("?limit=%d", limit))
    Assertions.assertThat(response.params.languages).contains("en", "de")
  }

  @Test
  fun getTranslations() {
    val project = dbPopulator.populate(generateUniqueString())
    val mvcResult = performAuthGet("/api/project/${project.id}/translations/en,de")
      .andExpect(status().isOk).andReturn()
    val result: Map<String, Any> = mvcResult.mapResponseTo()
    assertThat(result).containsKeys("en", "de")
  }

  @Test
  fun setTranslations() {
    val project = dbPopulator.createBase(generateUniqueString())
    val translationsMap = mapOf(Pair("en", "Hello"), Pair("de", "Hallo"))

    performAuthPost(
      "/api/project/${project.id}/translations",
      SetTranslationsWithKeyDto(
        "hello",
        translationsMap
      )
    ).andExpect(status().isOk)

    val fromService = translationService.getKeyTranslationsResult(
      project.id,
      PathDTO.fromFullPath("hello"),
      setOf("en", "de")
    )

    assertThat(fromService).isEqualTo(translationsMap)
  }

  @Test
  @ProjectApiKeyAuthTestMethod
  fun setTranslationsWithApiKey() {
    val translationsMap = mapOf(Pair("en", "Hello"), Pair("de", "Hallo"))

    performProjectAuthPost(
      "translations",
      SetTranslationsWithKeyDto(
        "hello",
        translationsMap
      )
    ).andExpect(status().isOk)

    val fromService = translationService.getKeyTranslationsResult(
      project.id,
      PathDTO.fromFullPath("hello"),
      setOf("en", "de")
    )

    assertThat(fromService).isEqualTo(translationsMap)
  }

  @Test
  @ProjectApiKeyAuthTestMethod(scopes = [ApiScope.TRANSLATIONS_EDIT])
  fun setTranslationsWithApiKeyForbidden() {
    val translationsMap = mapOf(Pair("en", "Hello"), Pair("de", "Hallo"))

    performProjectAuthPost(
      "translations",
      SetTranslationsWithKeyDto(
        "hello",
        translationsMap
      )
    ).andIsForbidden
  }

  @Test
  fun updateTranslations() {
    val project = dbPopulator.createBase(generateUniqueString())
    val translationsMap = mapOf(Pair("en", "Hello"), Pair("de", "Hallo"))
    keyService.create(project, SetTranslationsWithKeyDto(key = "hello", translations = translationsMap))

    val updatedTranslationsMap = mapOf(Pair("en", "Hello you!"), Pair("de", "Hallo dich!"))
    performAuthPut(
      "/api/project/${project.id}/translations",
      SetTranslationsWithKeyDto(
        "hello",
        updatedTranslationsMap
      )
    ).andExpect(status().isOk)

    val fromService = translationService.getKeyTranslationsResult(
      project.id,
      PathDTO.fromFullPath("hello"),
      setOf("en", "de")
    )

    assertThat(fromService).isEqualTo(updatedTranslationsMap)
  }

  @Test
  fun `denies access for user without the language permission`() {
    val testData = LanguagePermissionsTestData()
    testDataService.saveTestData(testData.root)
    loginAsUser(testData.enOnlyUser.username)
    performAuthPut(
      "/api/project/${testData.project.id}/translations",
      mapOf(
        "key" to "key",
        "translations" to mapOf(
          "de" to "de"
        )
      )
    ).andIsForbidden
  }

  private fun performGetDataForView(projectId: Long, queryString: String): ResultActions {
    return performAuthGet("/api/project/$projectId/translations/view$queryString")
  }
}
